lib/translator/write.cc

```diff
@@ -0,0 +1,217 @@
+  uint8_t pract, prchk;
+
+  switch (wrprotect) {
+    case 0b000:
+      pract = 1;
+      prchk = 0b000;
+      break;
+    case 0b001:
+      pract = 0;
+      prchk = 0b111;
+      break;
+    case 0b101:
+      pract = 0;
+      prchk = 0b111;
+      break;
+    case 0b010:
+      pract = 0;
+      prchk = 0b011;
+      break;
+    case 0b011:
+      pract = 0;
+      prchk = 0b00;
+      break;
+    case 0b100:
+      pract = 0;
+      prchk = 0b100;
+      break;
+    default:
+      // All other codes shall result in command termination with CHECK
+      // CONDITION status, ILLEGAL REQUEST sense key, and ILLEGAL FIELD IN CDB
+      // additional sense code.
+      DebugLog("Invalid WriteProtect Code for PRInfo.");
+      return StatusCode::kFailure;
+      break;
+  }
+
+  pr_info = prchk | (pract << 3);
+  return StatusCode::kSuccess;
+}
+
+// Converts transfer_length in units of logical blocks to units of pages
+uint32_t GetTransferLengthPages(uint16_t transfer_length, uint32_t page_size,
+                                uint32_t lba_size) {
+  uint64_t transfer_length_bytes = transfer_length * lba_size;
+  return transfer_length_bytes / page_size +
+         ((transfer_length_bytes % page_size == 0) ? 0 : 1);
+}
+
+// This function populates the nvme::GenericQueueEntryCmd object with fields
+// that are common to all Write Commands (6, 10, 12, 16) Refer to Section 5.7
+// (https://nvmexpress.org/wp-content/uploads/NVM_Express_-_SCSI_Translation_Reference-1_5_20150624_Gold.pdf)
+StatusCode LegacyWrite(nvme::GenericQueueEntryCmd& nvme_cmd,
+                       Allocation& allocation, uint32_t nsid,
+                       uint32_t transfer_length_pages) {
+  StatusCode status_code = allocation.SetPages(transfer_length_pages, 0);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  nvme_cmd = {.opc = static_cast<uint8_t>(nvme::NvmOpcode::kWrite),
+              .psdt = 0,  // prps are used
+              .nsid = nsid};
+
+  nvme_cmd.dptr.prp.prp1 = allocation.data_addr;
+
+  return status_code;
+}
+
+StatusCode Write(bool fua, uint8_t wrprotect, uint32_t transfer_length,
+                 nvme::GenericQueueEntryCmd& nvme_cmd, Allocation& allocation,
+                 uint32_t nsid, uint32_t page_size, uint32_t lba_size) {
+  if (transfer_length == 0) {
+    DebugLog("NVMe read command does not support transfering zero blocks\n");
+    return StatusCode::kNoTranslation;
+  }
+
+  transfer_length &= 0xffff;  // truncate to 16 bits
+  StatusCode status_code =
+      LegacyWrite(nvme_cmd, allocation, nsid,
+                  GetTransferLengthPages(transfer_length, page_size, lba_size));
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  uint8_t pr_info;
+  status_code = BuildPRInfo(wrprotect, pr_info);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+  nvme_cmd.cdw[2] = (transfer_length - 1 | pr_info << 26 | fua << 30);
```

Comment:
at this point, transfer length should be in host endian. and the whole expression should be wrapped in `htoll()`
```suggestion
  nvme_cmd.cdw[2] = htoll(transfer_length - 1 | pr_info << 26 | fua << 30);
```

Reply:
https://github.com/googleinterns/scsi2nvme/pull/128#discussion_r465369072